File: c/selectionsort.c
// Selection Sort

#include <stdio.h>


// Sort an array using selection sort
void sort(int *arr, int arrLen);

int main() {
    int arr[] = {5, 9, 4, 3, 2, 1};

    sort(arr, 6);

    for (int i = 0; i < 6; i++)
        printf("%d ", arr[i]);
    putchar('\n');

    return 0;
}

void sort(int *arr, int arrLen) {
    int maxInd, temp;

    // Repeat steps n - 1 times (worst case)
    for (int i = 0; i < arrLen - 1; ++i) {
        // Find the maximum item in unsorted part of the array
        maxInd = 0; 
        for (int j = 1; j < arrLen - i; ++j) {
            if (arr[j] > arr[maxInd])
                maxInd = j;
        }

        // Replace with last index in unsorted part of the array
        temp = arr[maxInd];
        arr[maxInd] = arr[arrLen - i - 1];
        arr[arrLen - i - 1] = temp;
    }

    return;
}

File: python/selectionsort.py
# Selection Sort

import random


def sort(arr):
    arrLen = len(arr)

    # Repeat step N - 1 times (worst case scenario)
    for i in range(arrLen - 1):
        unsortedIndex = arrLen - i - 1  # The index till which the array is unsorted (reduces every step)
        maxInd = 0
        # Find the maximum number in unsorted part of the array
        for i in range(1, unsortedIndex + 1):  # +1 to include the index
            if arr[i] > arr[maxInd]: maxInd = i
        
        # Swap maximum number with last index of unsorted part of the array
        arr[maxInd], arr[unsortedIndex] = arr[unsortedIndex], arr[maxInd]

    return

arr = list(range(1000))
random.shuffle(arr)
sort(arr)

print(arr)
